esp32: Update to use ESP-IDF v5.5.1.

dpgeorge opened the pull request:
### Summary

This PR updates the esp32 port to use IDF version 5.5.1.

For ESP32 code size is increased by +19984 bytes (comparing v1.26.1 to this PR, 1734416 -> 1754400 bytes).

Metrics for this PR, v5.4.2 vs v5.5.1:

**Note: IRAM usage is reduced by up to 19KB by config changes in this PR. Those savings aren't reflected in this table (as metrics script doesn't also swap before/after config, and table was generated before the changes were applied.**

| BOARD | BOARD_VARIANT | IDF Version | Binary Size | Static IRAM Size | Static DRAM Size |
|-------|---------------|-------------|-------------|------------------|------------------|
| ESP32_GENERIC |  | v5.4.2 | 1673054 | 117142 | 54012 |
|  |  | v5.5.1 | 1692651 (+19597/1.2%) | 120319 (+3177/2.7%) | 54020 (+8/0.0%) |
| ESP32_GENERIC | D2WD | v5.4.2 | 1380786 | 110878 | 53680 |
|  |  | v5.5.1 | 1393943 (+13157/1.0%) | 112847 (+1969/1.8%) | 53724 (+44/0.1%) |
| ESP32_GENERIC | SPIRAM | v5.4.2 | 1526066 | 116114 | 54148 |
|  |  | v5.5.1 | 1543223 (+17157/1.1%) | 118211 (+2097/1.8%) | 54208 (+60/0.1%) |
| ESP32_GENERIC_C2 |  | v5.4.2 | 1494376 | 40576 | 38508 |
|  |  | v5.5.1 | 1506236 (+11860/0.8%) | 43326 (+2750/6.8%) | 38316 (-192/-0.5%) |
| ESP32_GENERIC_C6 |  | v5.4.2 | 1842510 | 104730 | 45564 |
|  |  | v5.5.1 | 1862642 (+20132/1.1%) | 112264 (+7534/7.2%) | 45768 (+204/0.4%) |
| ESP32_GENERIC_S3 |  | v5.4.2 | 1669734 | 109770 | 49848 |
|  |  | v5.5.1 | 1685799 (+16065/1.0%) | 113455 (+3685/3.4%) | 49896 (+48/0.1%) |
| ESP32_GENERIC_S3 | SPIRAM_OCT | v5.4.2 | 1673190 | 112410 | 50668 |
|  |  | v5.5.1 | 1689255 (+16065/1.0%) | 116095 (+3685/3.3%) | 50716 (+48/0.1%) |


### Testing

(This section will be updated over time as more testing is done.)

I've run tests with the following firmware:
- ESP32_GENERIC
- ESP32_GENERIC_C3
- ESP32_GENERIC_C6
- ESP32_GENERIC_S3

Using the entire available test suite (everything in `tests/`).  There don't seem to be any regressions.

Note that the following tests fail, but not due to the updated IDF (see comments below for reasons):
- `ports/esp32/resolve_on_connect.py`
- `ports/esp32/check_err_str.py` -- when run via-mpy

### Trade-offs and Alternatives

As usual with these IDF updates, the only choice we really have is when to do it.  We could hold back and do it later, or do it now.  The benefit of doing it now is to get P4 and C5 support.

dpgeorge committed:
esp32: Update to use ESP-IDF v5.5.1.

Tested with ESP32_GENERIC, ESP32_GENERIC_C3, ESP32_GENERIC_C6 and
ESP32_GENERIC_S3 firmware, using the full test suite.  There do not seem to
be any regressions.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge commented:
Related:
- #17912
- #17951 
- #18063

DvdGiessen commented:
For what it's worth, I've been compiling against v5.5.1 (and v5.5

DvdGiessen commented:
Oh, I did run into one thing: I had to set one additional `sdkconfig` option to work around a known issue:

```diff
+# Workaround for bug in IDFv5.5-v5.5.1
+# See https://github.com/espressif/esp-idf/issues/17436#issuecomment-3197206868
+CONFIG_NEWLIB_NANO_FORMAT=n
+
```

Given they have it fixed I expect this will be resolved in the upcoming v5.5.2, but so far I've had to build locally with this extra option set.

robert-hh commented:
Which OS & Version do you use. When switching to v5.5.1, I kept the setting `CONFIG_NEWLIB_NANO_FORMAT=y` and could build esp32 firmware for various devices.

dpgeorge commented:
> Oh, I did run into one thing: I had to set one additional `sdkconfig` option to work around a known issue:
> ...
> CONFIG_NEWLIB_NANO_FORMAT=n

OK, thanks for that tip.  I'll have to think how to get that option working just for ESP32, because we currently explicitly enable it, to save code size.

dpgeorge commented:
> As a patch on can add `CONFIG_NEWLIB_NANO_FORMAT=n` to sdkconfig,spiram,

Thanks for the suggestion @robert-hh .  I've now done that in this PR.

@projectgus this is ready for you to run `metrics_esp32.py` on :rocket:

projectgus committed:
esp32/boards: Apply additional IRAM savings.

Saves approx 19KB on ESP32-C6 and 11KB on ESP32-C2, meaning IDF V5.5.1
builds now have more free RAM than V5.4.2.

This work was funded through GitHub Sponsors.

Signed-off-by: Angus Gratton <angus@redyak.com.au>

projectgus commented:
@dpgeorge I clawed back about 19.5KB of IRAM on C6, and 11KB on C2, by tweaking some config. The free RAM after this PR should now be higher than before.

Have ran `multi_net` tests on C2 and C6, and some basic unit tests on the C2. Let me know if you'd like me to test more boards.
